### ISSUE TITLE
Function calls on distributed tables

### ISSUE THREAD

**Issue by pm1000** [citusdata/citus#6102]
Hi all,

for the last few days I tried some tests with Citus. I noticed that Citus does not support standard SQL functions that query distributed tables (even if they don't take any parameters). I don't really see the overhead that stable inline functions create for Citus that they aren't supported by Citus.

The "best" workaround that I found so far is to use views instead of functions. That works for the functions without parameters but is there any effort to support SQL functions in the future or will this not be implemented? 

Thanks in advance:)

**Comment by onurctirtir** [comment]
Hey @pm1000,

Could you share the error message you've seen, as I couldn't follow what does the following mean:
> Citus does not support standard SQL functions that query distributed tables

```sql
create table dist (dist_col int, int_col int);
SELECT create_distributed_table('dist', 'dist_col');

CREATE OR REPLACE FUNCTION read_dist_table()
RETURNS setof int STABLE
LANGUAGE plpgsql
AS $function$
BEGIN
	return query select int_col from dist;
END;
$function$;

select read_dist_table();
┌─────────────────┐
│ read_dist_table │
├─────────────────┤
└─────────────────┘
(0 rows)
```

**Comment by pm1000** [comment]
Yes it works with plpgsql functions but I want to use normal sql functions that can be inlined when they are called.

Here is what I want to do:
`
begin;

	create table test1 (
		dist_col text,
		random_col_1 text,
		random_col_2 int
	);
	
	create table test2 (
		dist_col text,
		random_col_1 text,
		random_col_2 int,
		random_col_3 int,
		random_col_4 text
	);

	select * from create_distributed_table('test1', 'dist_col');
	select * from create_distributed_table('test2', 'dist_col');


	create or replace function public.get_test1()
	returns setof test1 
	language sql
	stable parallel safe as 
	$function$
		select * from test1
	$function$;
	
	create or replace function public.get_test2(_dist_col text, _random_col_1 text, _random_col_2 int)
	returns setof test2
	language sql 
	stable parallel safe as 
	$function$
		select * from test2 where dist_col = _dist_col and random_col_1 = _random_col_1 and random_col_2 = _random_col_2
	$function$;

commit;

`

This results in this error:
`SQL Error [XX000]: ERROR: cannot run function command because there was a parallel operation on a distributed table in the transaction
  Detail: When running command on/for a distributed function, Citus needs to perform all operations over a single connection per node to ensure consistency.
  Hint: Try re-running the transaction with "SET LOCAL citus.multi_shard_modify_mode TO 'sequential';"`

Running the same transaction with the suggested addition at least allows to create the functions but this does not really solve the problem.

If I try to make a normal query on one of the functions like that:
`
select * from get_test1();
`
Then this error appears:
`
SQL Error [0A000]: ERROR: cannot perform distributed planning on this query because parameterized queries for SQL functions referencing distributed tables are not supported
  Hint: Consider using PL/pgSQL functions instead.
`

Plpgsql functions work but they don't really fit my use case. Creating the functions as views works fine for functions without parameters but I need to have the functions with parameters as well.

I want to use the functions like that:
`
select * from get_test1() t1
left join get_test2(t1.dist_col, t1.random_col_1, t1.random_col_2) t2 on true;
`
I don't really see a reason why this is not supported to be honest because executing those queries "manually" works fine:
`select * from test1 t1
left join test2 t2 on (t1.dist_col = t2.dist_col and t1.random_col_1 = t1.random_col_1 and t1.random_col_2 = t2.random_col_2);
`
or
`
select * from test1 t1
left join test2 t2 on true 
where t1.dist_col = t2.dist_col and t1.random_col_1 = t1.random_col_1 and t1.random_col_2 = t2.random_col_2;
`

Feel free to ask me if something was unclear.

**Comment by fvannee** [comment]
It's also quite easy to trigger an assert in above example with:

```
select (select 1 from test1 limit 1) from get_test1();
```

```
/* GetRTEIdentity returns the identity assigned with AssignRTEIdentity. */
int
GetRTEIdentity(RangeTblEntry *rte)
{
	Assert(rte->rtekind == RTE_RELATION);
	Assert(rte->values_lists != NIL);    <-------- this is the assert that triggers
	Assert(IsA(rte->values_lists, IntList));
	Assert(list_length(rte->values_lists) == 2);

	return linitial_int(rte->values_lists);
}
```

Is there any plan in Citus to support inlineable functions? They are really useful as they can make user queries a lot simpler, while still providing all the benefits of being flexible. If you put functions in plpgsql you cannot efficiently use aggregates or where-clauses on it outside of the function. For example `select * from some_func(24) where col='abc'`, which, if not inlined, would require `some_func(24)` to execute fully and return all rows. The filter on `col` can only be applied afterwards. If it's inlined, it could determine that there's possibly a useful index on `col`.

**Comment by onurctirtir** [comment]
> Yes it works with plpgsql functions but I want to use normal sql functions that can be inlined when they are called.
> 
> Here is what I want to do:
> 
> ```
> begin;
> 
> 	create table test1 (
> 		dist_col text,
> 		random_col_1 text,
> 		random_col_2 int
> 	);
> 	
> 	create table test2 (
> 		dist_col text,
> 		random_col_1 text,
> 		random_col_2 int,
> 		random_col_3 int,
> 		random_col_4 text
> 	);
> 
> 	select * from create_distributed_table('test1', 'dist_col');
> 	select * from create_distributed_table('test2', 'dist_col');
> 
> 	create or replace function public.get_test1()
> 	returns setof test1 
> 	language sql
> 	stable parallel safe as 
> 	$function$
> 		select * from test1
> 	$function$;
> 
> 	create or replace function public.get_test2(_dist_col text, _random_col_1 text, _random_col_2 int)
> 	returns setof test2
> 	language sql 
> 	stable parallel safe as 
> 	$function$
> 		select * from test2 where dist_col = _dist_col and random_col_1 = _random_col_1 and random_col_2 = _random_col_2
> 	$function$;
> 
> commit;
> ```
> 
> This results in this error: `SQL Error [XX000]: ERROR: cannot run function command because there was a parallel operation on a distributed table in the transaction Detail: When running command on/for a distributed function, Citus needs to perform all operations over a single connection per node to ensure consistency. Hint: Try re-running the transaction with "SET LOCAL citus.multi_shard_modify_mode TO 'sequential';"`
> 
> Running the same transaction with the suggested addition at least allows to create the functions but this does not really solve the problem.
> 
> If I try to make a normal query on one of the functions like that:
> 
> `select * from get_test1(); ` Then this error appears: `SQL Error [0A000]: ERROR: cannot perform distributed planning on this query because parameterized queries for SQL functions referencing distributed tables are not supported Hint: Consider using PL/pgSQL functions instead.`
> 
> Plpgsql functions work but they don't really fit my use case. Creating the functions as views works fine for functions without parameters but I need to have the functions with parameters as well.
> 
> I want to use the functions like that: `select * from get_test1() t1 left join get_test2(t1.dist_col, t1.random_col_1, t1.random_col_2) t2 on true;` I don't really see a reason why this is not supported to be honest because executing those queries "manually" works fine: `select * from test1 t1 left join test2 t2 on (t1.dist_col = t2.dist_col and t1.random_col_1 = t1.random_col_1 and t1.random_col_2 = t2.random_col_2); ` or `select * from test1 t1 left join test2 t2 on true where t1.dist_col = t2.dist_col and t1.random_col_1 = t1.random_col_1 and t1.random_col_2 = t2.random_col_2;`
> 
> Feel free to ask me if something was unclear.

Thank you for all the details you shared, let me label this as a feature request.

**Comment by onurctirtir** [comment]
> select (select 1 from test1 limit 1) from get_test1();

And will open a separate issue for this assertion failure.

**Comment by fvannee** [comment]
@onurctirtir do you have a rough idea what would be required to implement this feature? I had a quick look at the code, but it seems that it's not trivial due to the way that Citus intercepts the queries before the function inlining takes place. Would appreciate any insights you might have.

**Comment by fvannee** [comment]
Any idea on this? I'd be interested in giving implementing this a try myself, as not having inlinable sql functions is one of the main limitations of Citus I see right now. We're using these very often to abstract complex queries from users and inlineable is a necessity for having the functions be nicely composable. However from looking at code, it seems like it's a rather hard limitation on the way Citus currently implements parsing the query.
